lib/dnscache.c

```diff
@@ -166,37 +144,92 @@ dns_cache_fill_key(DNSCacheKey *key, gint family, void *addr)
 }
 
 static void
-dns_cache_cleanup_persistent_hosts(void)
+dns_cache_store(DNSCache *self, gboolean persistent, gint family, void *addr, const gchar *hostname, gboolean positive)
 {
-  while (persist_first.next != &persist_last)
+  DNSCacheEntry *entry;
+  guint hash_size;
+
+  entry = g_new(DNSCacheEntry, 1);
+
+  dns_cache_fill_key(&entry->key, family, addr);
+  entry->hostname = g_strdup(hostname);
+  entry->hostname_len = strlen(hostname);
+  entry->positive = positive;
+  INIT_IV_LIST_HEAD(&entry->list);
+  if (!persistent)
     {
-      g_hash_table_remove(cache, &persist_first.next->key);
-      dns_cache_persistent_count--;
+      entry->resolved = cached_g_current_time_sec();
+      iv_list_add(&self->cache_list, &entry->list);
+    }
+  else
+    {
+      entry->resolved = 0;
+      iv_list_add(&self->persist_list, &entry->list);
+    }
+  hash_size = g_hash_table_size(self->cache);
+  g_hash_table_replace(self->cache, &entry->key, entry);
+
+  if (persistent && hash_size != g_hash_table_size(self->cache))
+    self->persistent_count++;
+
+  /* persistent elements are not counted */
+  if ((gint) (g_hash_table_size(self->cache) - self->persistent_count) > self->options->cache_size)
+    {
+      DNSCacheEntry *entry = iv_list_entry(&self->cache_list, DNSCacheEntry, list);
```

Comment:
This variable (entry) should be rename to something else. That's why the travis failed using -Wshadow. What do you think about entry_to_remove?